src/main/kotlin/ee/carlrobert/codegpt/codecompletions/InfillPromptTemplate.kt

```diff
@@ -1,42 +1,97 @@
 package ee.carlrobert.codegpt.codecompletions
 
+import com.intellij.openapi.vfs.readText
+
 enum class InfillPromptTemplate(val label: String, val stopTokens: List<String>?) {
 
     OPENAI("OpenAI", null) {
-        override fun buildPrompt(prefix: String, suffix: String): String {
-            return "<|fim_prefix|> $prefix <|fim_suffix|>$suffix <|fim_middle|>"
+        override fun buildPrompt(infillDetails: InfillRequestDetails): String {
+            return "<|fim_prefix|> ${infillDetails.prefix} <|fim_suffix|>${infillDetails.suffix} <|fim_middle|>"
         }
     },
     CODE_LLAMA("Code Llama", listOf("<EOT>")) {
-        override fun buildPrompt(prefix: String, suffix: String): String {
-            return "<PRE> $prefix <SUF>$suffix <MID>"
+        override fun buildPrompt(infillDetails: InfillRequestDetails): String {
+            return "<PRE> ${infillDetails.prefix} <SUF>${infillDetails.suffix} <MID>"
         }
     },
     CODE_GEMMA(
         "CodeGemma Instruct",
         listOf("<|file_separator|>", "<|fim_prefix|>", "<|fim_suffix|>", "<|fim_middle|>", "<eos>")
     ) {
-        override fun buildPrompt(prefix: String, suffix: String): String {
-            return "<|fim_prefix|>$prefix<|fim_suffix|>$suffix<|fim_middle|>"
+        override fun buildPrompt(infillDetails: InfillRequestDetails): String {
+            // see https://huggingface.co/google/codegemma-7b#for-code-completion
+            val infillPrompt =
+                "<|fim_prefix|>${infillDetails.prefix}<|fim_suffix|>${infillDetails.suffix}<|fim_middle|>"
+            return if (infillDetails.context == null || infillDetails.context.contextFiles.isNullOrEmpty()) {
+                infillPrompt
+            } else {
+                infillDetails.context.contextFiles.map {
+                    "<|file_separator|>${it.path} \n" +
+                            it.readText()
```

Comment:
Do you mean to if e.g. there is a method call to [StringUtils.capitalize](https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/StringUtils.html#capitalize-java.lang.String-) that only the source code of the method should be included, not its entire class?
Yes that would be possible, although I am not sure if the models will get confused by that, because from what I read during training they always provided the **entire** source code file, not just "relevant" parts (for the models with repository-level completion tokens). But I would also prefer just to send the relevant parts obviously.

Reply:
There is a method [ClsMethodImpl.sourceMirrorMethod](https://github.com/JetBrains/intellij-community/blob/57f19ccb362b9dedc4fb374c58cf2bfd756a3b3a/java/java-psi-impl/src/com/intellij/psi/impl/compiled/ClsMethodImpl.java#L247) which seems to return the source code of the method as a `PsiElement`. If you did not yet use the  "Download sources" action it returns `null`, but if you did it return this, e.g. for [StringUtils.capitalize](https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/StringUtils.html#capitalize-java.lang.String-):

<details>
  <summary>ClsMethodImpl.sourceMirrorMethod.text</summary>
  
  ```java
/**
     * <p>Capitalizes a String changing the first character to title case as
     * per {@link Character#toTitleCase(int)}. No other characters are changed.</p>
     *
     * <p>For a word based algorithm, see {@link org.apache.commons.lang3.text.WordUtils#capitalize(String)}.
     * A {@code null} input String returns {@code null}.</p>
     *
     * <pre>
     * StringUtils.capitalize(null)  = null
     * StringUtils.capitalize("")    = ""
     * StringUtils.capitalize("cat") = "Cat"
     * StringUtils.capitalize("cAt") = "CAt"
     * StringUtils.capitalize("'cat'") = "'cat'"
     * </pre>
     *
     * @param str the String to capitalize, may be null
     * @return the capitalized String, {@code null} if null String input
     * @see org.apache.commons.lang3.text.WordUtils#capitalize(String)
     * @see #uncapitalize(String)
     * @since 2.0
     */
    public static String capitalize(final String str) {
        final int strLen = length(str);
        if (strLen == 0) {
            return str;
        }

        final int firstCodepoint = str.codePointAt(0);
        final int newCodePoint = Character.toTitleCase(firstCodepoint);
        if (firstCodepoint == newCodePoint) {
            // already capitalized
            return str;
        }

        final int[] newCodePoints = new int[strLen]; // cannot be longer than the char array
        int outOffset = 0;
        newCodePoints[outOffset++] = newCodePoint; // copy the first codepoint
        for (int inOffset = Character.charCount(firstCodepoint); inOffset < strLen; ) {
            final int codepoint = str.codePointAt(inOffset);
            newCodePoints[outOffset++] = codepoint; // copy the remaining ones
            inOffset += Character.charCount(codepoint);
         }
        return new String(newCodePoints, 0, outOffset);
    }
```
</details>

Reply:
> Do you mean to if e.g. there is a method call to StringUtils.capitalize that only the source code of the method should be included, not its entire class?

Exactly!

> Yes that would be possible, although I am not sure if the models will get confused by that, because from what I read during training they always provided the entire source code file, not just "relevant" parts (for the models with repository-level completion tokens). But I would also prefer just to send the relevant parts obviously.

I tried sending snippets of code before the prefix of a cursor, and `gpt-3.5-instruct` didn't seem to struggle with it, given that the entire context was syntactically incorrect. However, it may be different per model.

Reply:
Okay, I guess we can try out some simple template like:
```
# file_x.py:
<code snippet>

# file_y.py:
<code snippet>

# file_z.py:
<|fim_prefix|> ${infillDetails.prefix} <|fim_suffix|>${infillDetails.suffix} <|fim_middle|>
```

Hopefully at some point there will be a successor of the `gpt-3.5-instruct` model with better support for multi-level completions

Reply:
> I think in general the decompiled code should still be usable for the LLMs, although obviously the actual source code would be better.

I think it's fine, let's keep it for now.

> A major concern is how we limit the prompt/context size, especially when Classes are referenced, can we somehow evaluate what parts of the class is relevant? Do we limit the amount of referenced classes/methods?

This is difficult. I think we need to set some maximum token limit boundaries, and even without the class references, some of the context may be left out or even worse, cut out. So, there needs to exist some kind of reranking mechanism to filter out irrelevant data. For now, could we just filter out the class elements?

> Also you can see that since in the FIM prompt itself there is only the method getMoviesSortedByPlaytime, the class field movieService is not present anymore, so the LLM would probably give a faulty/unuseful completion.

Would it be possible to provide the class name or file path from which the `getMoviesSortedByPlaytime` method is coming? For example, instead of passing in the list of `PsiTargets`, you could forward a mapping where the key points to class details and the value to its methods/functions?